File: Comprehensive_Data_Operations_Simulation/Q2. Range Frequency Queries/Q2_RangeFrequencyQueries.py
class RangeFreqQuery(object):                     # define the class

    def __init__(self, arr):
        """ :type arr: List[int] """
        
        self.map = {}                              # store value → list of positions

        for i, val in enumerate(arr):              # loop through array with index
            if val not in self.map:                # if value not seen before
                self.map[val] = []                 # create a new list
            self.map[val].append(i)                # add index of the value

    def query(self, left, right, value):
        """
        :type left: int
        :type right: int
        :type value: int
        :rtype: int
        """
        
        if value not in self.map:                  # if value not present in array
            return 0                               # return 0

        arr = self.map[value]                      # get list of indices for value

        l = self.lower_bound(arr, left);           # find first index >= left
        r = self.upper_bound(arr, right);          # find first index > right

        return r - l                               # count indices inside the range

    def lower_bound(self, arr, target):            # binary search: first >= target
        l, r = 0, len(arr)                         # search range
        while l < r:                               # loop until bounds meet
            mid = (l + r) // 2                     # middle index
            if arr[mid] >= target:                 # if mid value is big enough
                r = mid                            # move left
            else:
                l = mid + 1                        # move right
        return l                                   # return position

    def upper_bound(self, arr, target):            # binary search: first > target
        l, r = 0, len(arr)                         # search range
        while l < r:                               # loop until bounds meet
            mid = (l + r) // 2                     # middle index
            if arr[mid] > target:                  # if mid value is greater
                r = mid                            # move left
            else:
                l = mid + 1                        # move right
        return l                                   # return position

File: Comprehensive_Data_Operations_Simulation/Q2. Range Frequency Queries/Q2_RangeFrequencyQueries.java
class RangeFreqQuery {                           // define the RangeFreqQuery class

    private Map<Integer, List<Integer>> map;     // map to store value → list of indices

    public RangeFreqQuery(int[] arr) {           // constructor taking input array
        map = new HashMap<>();                   // create the HashMap

        for (int i = 0; i < arr.length; i++) {   // loop through all elements
            map.computeIfAbsent(arr[i],          // if value not in map, create list
                k -> new ArrayList<>())
                .add(i);                         // add index of value
        }
    }
    
    public int query(int left, int right, int value) {   // method to answer queries
        if (!map.containsKey(value)) return 0;           // if value not in array, return 0

        List<Integer> list = map.get(value);             // get index list of the value

        int l = lowerBound(list, left);                  // find first index >= left
        int r = upperBound(list, right);                 // find first index > right

        return r - l;                                    // count indices within range
    }

    private int lowerBound(List<Integer> list, int target) { // binary search lower bound
        int l = 0, r = list.size();                          // search range
        while (l < r) {                                      // continue while valid
            int mid = l + (r - l) / 2;                       // find middle
            if (list.get(mid) >= target) r = mid;            // move left if >= target
            else l = mid + 1;                                // else move right
        }
        return l;                                            // return first ≥ target
    }

    private int upperBound(List<Integer> list, int target) { // binary search upper bound
        int l = 0, r = list.size();                          // search range
        while (l < r) {                                      // continue while valid
            int mid = l + (r - l) / 2;                       // find middle
            if (list.get(mid) > target) r = mid;             // move left if > target
            else l = mid + 1;                                // else move right
        }
        return l;                                            // return first > target
    }
}
